File: algorithm/leetcodeObjectiveC/palindrome-permutation-II.m
// 267. Palindrome Permutation II
// Difficulty : Medium

// Given a string s, return all the palindromic permutations (without duplicates) of it.
// Return an empty list if no palindromic permutation could be form.

// For example:

// Given s = "aabb", return ["abba", "baab"].

// Given s = "abc", return [].

// Hint:
// If a palindromic permutation exists, we just need to generate the first half of the string.
// To generate all distinct permutations of a (half of) string, use a similar approach from:
// Permutations II or Next Permutation.

// Time:  O(n * n!)
// Space: O(n)

#import <Foundation/Foundation.h>

void permuteHelper(NSString* chars, BOOL *visited, NSString* path, NSMutableArray** ans) {
  if ([chars length] == [path length]) {
    [*ans addObject:[path mutableCopy]];
    return;
  }
  
  for (NSInteger i = 0; i < [chars length]; i++) {
    if (!visited[i]) {
      visited[i] = YES;
      permuteHelper(chars, visited, [path stringByAppendingFormat:@"%c", [chars characterAtIndex:i]], ans);
      visited[i] = NO;
    }
  }
}

NSArray* generatePalindromes(NSString* s) {
  if (![s length]) {
    return @[];
  }
  NSMutableDictionary* cnt = @{}.mutableCopy;
  for (int i = 0; i < s.length; i++) {
    cnt[[s substringWithRange:NSMakeRange(i, 1)]] = @([cnt[[s substringWithRange:NSMakeRange(i, 1)]] intValue] + 1);
  }
  NSMutableString* mid = @"".mutableCopy;
  NSMutableString* chars = @"".mutableCopy;
  for (id key in cnt) {
    if ([cnt[key] intValue] % 2) {
      if (!mid.length) {
        [mid appendString:key];
      } else {  // The odd occurance of chars in palindrome is at most once.
        return @[];
      }
    }
    chars = [[chars stringByPaddingToLength:chars.length + [cnt[key] intValue] / 2 withString:key startingAtIndex:0] mutableCopy];
  }
  // permute
  BOOL visited[chars.length];
  memset(visited, NO, sizeof(BOOL) * (chars.length));
  NSMutableArray* ans = @[].mutableCopy;
  permuteHelper(chars, visited, @"", &ans);
  // append mid + reverseChars
  for (int i = 0; i < [ans count]; i++) {
    NSString* reverseChars = [[[[ans[i] componentsSeparatedByString:@""] reverseObjectEnumerator] allObjects] componentsJoinedByString:@""];
    [ans[i] appendFormat:@"%@%@", mid, reverseChars];
  }
  return ans;
}

File: algorithm/cpp/palindrome-permutation-ii.cpp
// 267. Palindrome Permutation II
// Difficulty : Medium 

// Given a string s, return all the palindromic permutations (without duplicates) of it.
// Return an empty list if no palindromic permutation could be form.

// For example:

// Given s = "aabb", return ["abba", "baab"].

// Given s = "abc", return [].

// Hint:
// If a palindromic permutation exists, we just need to generate the first half of the string.
// To generate all distinct permutations of a (half of) string, use a similar approach from: 
// Permutations II or Next Permutation.

// Time:  O(n * n!)
// Space: O(n)

class Solution {
public:
  vector<string> generatePalindromes(string s) {
    if (s.empty()) {
      return {};
    }

    unordered_map<char, int> cnt;
    for (const auto& c : s) {
      ++cnt[c];
    }

    string mid, chars;
    for (const auto& kvp : cnt) {
      if (kvp.second % 2) {
        if (mid.empty()) {
          mid.push_back(kvp.first);
        } else {
          return {};
        }
      }
      chars.append(kvp.second / 2, kvp.first);
    }
    return permuteUnique(mid, chars);
  }
  
  vector<string> permuteUnique(const string& mid, string& chars) {
    vector<string> result;
    sort(chars.begin(), chars.end());
    do {
      string reverse_chars(chars.crbegin(), chars.crend());
      result.emplace_back(chars + mid + reverse_chars);
    } while (next_permutation(chars.begin(), chars.end()));
    return result;
  }
};